Add tests for time conversions in tools package

markcampanelli opened the pull request:
<!-- Thank you for your contribution! The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items. Feel free to remove checklist items that are not relevant to your change. -->

 - [x] Closes #2340
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing/index.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

<!-- Brief description of the problem and proposed solution (if not already fully described in the issue linked to above): -->

markcampanelli commented:
I realize no one planned/asked for this, but I think it's a good addition and ready to be reviewed by a maintainer. Increases test coverage by a whopping 0.07%!

cwhanse commented on pvlib/tests/test_tools.py:
```diff
@@ -144,3 +147,192 @@ def test_get_pandas_index(args, args_idx):
+    'input,expected',
+    [
+        (
+            {
```
The `dict` here appears only to be so that the function call can be double-splatted `(**input)`. I'm interested in feedback from others: is this parameterization style easier or more difficult, as a reviewer? It's not the style used for other tests are parameterized.

kandersolar replied:
My 2 cents: a more compact representation of the changing values is helpful for reviewing, and is worth a bit of processing inside the test function itself.  For example, it seems like the following:
```python
(
    {
        "time": pd.DatetimeIndex(
            ["1974-06-22T18:30:15"],
            tz=ZoneInfo("Etc/GMT+5"),
        ),
        "location": location.Location(
            43.19262774396091, -77.58782907414867, tz="Etc/GMT+5"
        )
    },
    pd.DatetimeIndex(["1974-06-22T23:30:15"], tz=ZoneInfo("UTC")),
),
(
    {
        "time": pd.DatetimeIndex(["1974-06-22T18:30:15"]),
        "location": location.Location(
            43.19262774396091, -77.58782907414867, tz="Etc/GMT+5"
        )
    },
    pd.DatetimeIndex(["1974-06-22T23:30:15"], tz=ZoneInfo("UTC")),
),
```

could be replaced with something like this: 
```python        
("1974-06-22T18:30:15", "Etc/GMT+5", 43, -77, "Etc/GMT+5", "1974-06-22T23:30:15+00:00"),
("1974-06-22T18:30:15", None, 43, -77, "Etc/GMT+5", "1974-06-22T23:30:15+00:00"),
```
And the test function could then contain the necessary calls to `Location` and `pd.DatetimeIndex` to construct the parameters themselves.

I think the latter format makes it easier to see what changes between each set of parameters.  I found my eyes having to flick back and forth and subsequently getting lost with the former format.

markcampanelli replied:
In the end I think this was simplified to
```python
@pytest.mark.parametrize(
    'tz,tz_expected', [
        pytest.param('UTC', 'UTC'),
        pytest.param('Etc/GMT+5', 'Etc/GMT+5'),
        pytest.param('US/Mountain', 'US/Mountain'),
        pytest.param('America/Phoenix', 'America/Phoenix'),
        pytest.param('Asia/Kathmandu', 'Asia/Kathmandu'),
        pytest.param('Asia/Yangon', 'Asia/Yangon'),
        pytest.param(datetime.timezone.utc, 'UTC'),
        pytest.param(zoneinfo.ZoneInfo('Etc/GMT-7'), 'Etc/GMT-7'),
        pytest.param(pytz.timezone('US/Arizona'), 'US/Arizona'),
        pytest.param(-6, 'Etc/GMT+6'),
        pytest.param(-11.0, 'Etc/GMT+11'),
        pytest.param(12, 'Etc/GMT-12'),
    ],
)
...
```

cwhanse reviewed (commented):
Thanks @markcampanelli 

The tests expose a weakness in `localize_to_utc`: what happens if `location` has a timezone defined as an offset? [That is permissible](https://github.com/pvlib/pvlib-python/blob/9fb2eb3aa7984c6283252e53e3052fe2c2cee90e/pvlib/location.py#L77), and if done produces an unhelpful message:

```
    if zone.upper() == 'UTC':

AttributeError: 'int' object has no attribute 'upper'
```

markcampanelli commented:
> Thanks @markcampanelli
> 
> The tests expose a weakness in `localize_to_utc`: what happens if `location` has a timezone defined as an offset? [That is permissible](https://github.com/pvlib/pvlib-python/blob/9fb2eb3aa7984c6283252e53e3052fe2c2cee90e/pvlib/location.py#L77), and if done produces an unhelpful message:
> 
> ```
>     if zone.upper() == 'UTC':
> 
> AttributeError: 'int' object has no attribute 'upper'
> ```

Good catch! Lemme see what I might do about fixing that here (if it's not too much scope creep).

BTW: I'm pretty sure this is the sort of issue that solved (hopefully!) by simplifying the timezone representation in `pvlib.location.Location` in https://github.com/pvlib/pvlib-python/pull/2343

markcampanelli commented:
@cwhanse @kandersolar I addressed the bug Cliff spotted, and updated a `Location` test for that.

Due to the high repetition, I decided not to use test parameterization at all for the new `tools` tests. Same coverage.

Should be ready for another look.

cwhanse commented on pvlib/location.py:
```diff
@@ -72,11 +72,12 @@ def __init__(self, latitude, longitude, tz='UTC', altitude=None,
-            self.pytz = tz
+            self.pytz = pytz.timezone(tz.zone)
         elif isinstance(tz, (int, float)):
-            self.tz = tz
```
There was a reason for storing the number here, but it's probably traceable all the way back to Matlab and no longer relevant.

markcampanelli replied:
I am open to only storing the `pytz`, but we might provide a read-only `tz` property that prints the IANA `str`. (See my other response, where I say that I also don't like the duplication of state.)

cwhanse commented on pvlib/tools.py:
```diff
@@ -119,21 +119,21 @@ def atand(number):
     time : datetime.datetime, pandas.DatetimeIndex,
            or pandas.Series/DataFrame with a DatetimeIndex.
-    location : pvlib.Location object
+    location : pvlib.Location object (unused if time is localized)
```
```suggestion
    location : pvlib.Location object (unused if ``time`` is localized)
```

cwhanse commented on pvlib/location.py:
```diff
@@ -72,11 +72,12 @@ def __init__(self, latitude, longitude, tz='UTC', altitude=None,
         elif isinstance(tz, (int, float)):
-            self.tz = tz
-            self.pytz = pytz.FixedOffset(tz*60)
+            self.tz = f"Etc/GMT{int(-tz):+d}"
```
I'm inclined to ask to revert this change. The docstring says `tz` can be a float. There are places that have a non-integer offset from UTC, e.g., the Marquesas are [UTC−09:30](https://en.wikipedia.org/wiki/UTC%E2%88%9209:30)[UTC−09:30](https://en.wikipedia.org/wiki/UTC%E2%88%9209:30). `localize_to_utc` is changed in this PR to examine `Location.pytz` rather than `Location.tz`, so the error we discussed won't occur.

markcampanelli replied:
There is a point in the Location docstring that says this:

>     Location objects have two timezone attributes:

        * ``tz`` is a IANA timezone string.
        * ``pytz`` is a pytz timezone object.

While I appreciate my changes may break some things, I think this existing docstring suggests a much more sane approach. (Indeed, in the sequel using `zoneinfo`, I propose using one, and only one, internal representation of the timezone, instead of two, potentially with helper functions s to update this internal representation using int/float offsets, etc..)

cwhanse replied:
I'm not saying revert from storing a string, but revert the `int` portion. If `tz` is truncated to an integer, how are timezones such as UTC+05:30 represented?

I'm open to doing away with accepting `int` or `float` for `tz`.

markcampanelli replied:
@cwhanse Sorry I misunderstood your meaning. I was aware that some locations are on half-hour offsets and such, but what is unclear to me is if/how that is represented in `pytz` (or `zoneinfo` for that matter). I do not see any fractional offsets in `zoneinfo.available_timezones()` and this also does not work for me: `pytz.timezone("Etc/GMT+5.5")`.

I must be missing something here. Does something non-standard have to be done to support fractional offsets?

markcampanelli replied:
> I'm open to doing away with accepting `int` or `float` for `tz`.

Hi again @cwhanse. It looks like the standard "non-integral-hour" offsets are only supported in the IANA standard through place names such as "Asia/Kathmandu" and "Asia/Yangon", but _not_ through `Etc/GMT+X.Y` style offsets.

Given that, I think the most trouble-free thing to do is to remove the `int` and `float` options and be sure to discuss this issue in the documentation tutorial. (Also, the only supported `tz` strings would be the IANA ones.) Sounds like you are amenable to this.

Keeping both fields `tz` (only as a string) and `pytz` (as a `pytz` timezone type), then I think I should add setters that change one when the other is changed after the Location object has already been already initialized.

cwhanse replied:
A proposal for the minimum impact on user code: do away with accepting `float` for the parameter `tz`, and keep your translation to a string offset. `Location.tz` is meant to be a string, so the mistake was setting `self.tz = tz` when `tz` is an integer or float.

For a future major version, we can discuss the merits of having only one of `tz` or `pytz`, of keeping both, and of transitioning away from `pytz`. I don't think we want to take steps in that direction in this minor PR that adds tests.

cwhanse replied:
>  I think the most trouble-free thing to do is to remove the int and float options 

I'm in favor of this change. @pvlib/pvlib-maintainer @pvlib/pvlib-triage your opinions?

adriesse replied:
No opinion.

markcampanelli replied:
Ok. I have added support back for int and non-fractional float being passed to `tz` in the `Location` constructor. I also test that `pytz` and `zoneinfo` timezones are likewise supported for initialization, and I beefed up the tests even more.

Note that the `pytz` Location attribute is now the single source of truth for the time zone in a `Location` object (which I recommend updating to a standard-library `timezone.TimeZone` object in #2342). `self.tz` simply returns the stringified `self.pytz`, but one can still directly set `self.tz` with any `tz` value supported by the `Location` intializer. Also, the `tz` and `pytz` attributes are kept in sync now, as changing one updates the other.

Ready for another review.

markcampanelli commented:
The two test failures appear to be unrelated to my changes.

cwhanse commented on pvlib/location.py:
```diff
@@ -38,12 +40,16 @@ class Location:
+        from the pytz and zoneinfo packages), default 'UTC'.
+        See http://en.wikipedia.org/wiki/List_of_tz_database_time_zones for a
+        list of valid name strings for IANA time zones.
+        ints and floats must be non-fractional N-hour offsets from UTC, which
```
```suggestion
        ints and floats must be whole number hour offsets from UTC, which
```

cwhanse commented on pvlib/location.py:
```diff
@@ -59,38 +65,54 @@ class Location:
+        elif isinstance(tz_, float):
+            if tz_ % 1 != 0:
+                raise TypeError(
+                    "floating point tz does not have zero fractional part: "
+                    f"{tz_}"
```
```suggestion
                    "floating point tz has non-zero fractional part: "
                    f"{tz_}. Only whole number offsets are supported."
```

cwhanse commented on pvlib/location.py:
```diff
@@ -59,38 +65,54 @@ class Location:
+        else:
+            raise TypeError(
+                f"invalid tz specification: {tz_}, must be an IANA time zone "
+                "string, a non-fractional int/float UTC offset, or a "
```
```suggestion
                "string, a whole number int/float UTC offset, or a "
```

cwhanse reviewed (commented):
Do you think we should switch and regard `tz` as the source of truth, and `pytz` as the optional attribute? With an eye toward replacing `pytz` with python's `timezone`.

markcampanelli commented:
> Do you think we should switch and regard `tz` as the source of truth, and `pytz` as the optional attribute? With an eye toward replacing `pytz` with python's `timezone`.

@cwhanse Ok, so we make the single source of truth the standardized IANA-supported string in the `tz` attribute, and make a `pytz` getter that converts `tz` to the `pytz` time zone using `return pytz.timezone(self.tz))`, and the setter that stringifies the provided `pytz` timezone into the `tz` string. I like this idea.

Question: Would it be too disruptive at this point to make the (newly implemented) `pytz` attribute "read only"? That is, remove the setter mentioned above, so that one can only change the time zone on an existing `Location` object by setting the `tz` attribute (with same interface as the `Location` initializer). I would much prefer this simplification too.

cwhanse commented:
> make the (newly implemented) pytz attribute "read only"? 

I don't think we do this. Since pytz is in maintenance, it seems reasonable to expect to replace pytz with timezone (as you anticipate). At that time, the pytz attribute can be removed. I don't see the value in having both the timezone object and a string as attributes, since one can print the object and get the string representation.

markcampanelli commented:
I decided to store the single source of truth about a `Location`'s time zone as a psuedo-private `zoneinfo.ZoneInfo` object. However, the time zone is still updated (only) using the `tz` attribute, and so that interface is unchanged (and now very consistent for new vs. existing Location objects). However, I did make the `pytz` attribute read only, which avoids the synchronization issue with the `tz` attribute. I also added a read-only `zoneinfo` attribute that newly allows uses to get the time zone as the standard library's `zoneinfo.ZoneInfo` object (which pandas is apparently moving to in 3.0). I updated the Location's docstring accordingly. This new foundation paves the ways for #2342, but with minimal current disruption.

I also made all of @cwhanse's recommended changes.

If this implementation and the removal of directly setting the `pytz` attribute is ok by everyone, then I would like to add a deprecation message to the `pytz` attribute that says zoneinfo will be the official time-zone package going forward. I also still need to update the `whatsnew` and verify that the new documentation renders properly.

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -26,5 +35,4 @@ Requirements
 ~~~~~~~~~~~~
-
-
+* Mark Campanellli (:ghuser:`markcampanelli`)
```
```suggestion
* Mark Campanelli (:ghuser:`markcampanelli`)
```

cwhanse reviewed (commented):
@markcampanelli let's keep this PR to the scope described in the title. Importing and adding the `zoneinfo` attribute feels like half of the journey to shifting from `pytz` to `zoneinfo` and I think that should be done in its own PR.

Same for fixing asv - do that in a separate PR.

markcampanelli commented:
@pvlib/pvlib-maintainer @pvlib/pvlib-triage Can someone review my `asv` CI fix here?

If this looks ok, then should I move the fix to a separate PR for better visibility?

markcampanelli commented:
> @pvlib/pvlib-maintainer @pvlib/pvlib-triage Can someone review my `asv` CI fix here?
> 
> If this looks ok, then should I move the fix to a separate PR for better visibility?

Moved to #2352, which now should go in before this.

cwhanse commented on pvlib/tools.py:
```diff
@@ -119,21 +120,21 @@ def atand(number):
 def localize_to_utc(time, location):
     """
-    Converts or localizes a time series to UTC.
+    Converts time to UTC, localizing if necessary using location.
```
```suggestion
    Converts ``time`` to UTC, localizing if necessary using location.
```

markcampanelli commented:
I'd like to say that I'm done with this PR and it's good to go content wise. I will do my best to try to check the generated API documentation for any formatting issues.

To clarify where things ended up with fixing the issues found in `Location` timezones:

1. Internally, a `Location` object's timezone is stored "psuedo-privately" as a `zoneinfo.ZoneInfo` object. (Now a single source of truth.)
2. The `Location` class's initializer interface should be compatible with existing code.
3. The `tz` attribute is now always an IANA-suppored string (as advertised) that can be updated safely after `Location` object initialization.
4. The `pytz` attribute is still supported for backwards compatibility with code that uses `pytz` timezones, but it is read only. One must (consistently) set the `tz` attribute to change a `Location`'s timezone, and setting this attribute still accepts a `pytz`-style timezone (same interface as the object initializer). However, `tz` will always read as a string.

markcampanelli commented:
I see that the documentation for the PR's code is generated in CI. Nice touch! Checking it now...

markcampanelli commented:
@pvlib/pvlib-maintainer Not sure if anyone knows how to document all the attributes in an object, not just those with a getter and setter as is apparently happening with `Location` here.

cwhanse commented on pvlib/location.py:
```diff
@@ -18,13 +19,17 @@
+        * ``tz`` is an IANA time-zone string.
+        * ``pytz`` is a pytz-based time-zone object (read only).
+
+    As with Location-object initialization, use the ``tz`` attribute to update
+    the Location's object's time zone after instantiation, and the read-only
+    ``pytz`` attribute will stay in sync with any changes made using ``tz``.
```
```suggestion
    The read-only ``pytz`` attribute will stay in sync with any changes made using ``tz``.
```

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -12,12 +12,24 @@ Enhancements
 
+Bug Fixes
+~~~~~~~~~
+* Ensure proper tz and pytz types in pvlib.location.Location. tz becomes the
+  single source of time-zone truth, is the single time-zone setter interface,
+  and the getter consistently returns an IANA string. datetime.tzinfo
+  subclasses are fully supported in tz setter. pytz attribute becomes read
+  only. (:issue:`2340`, :pull:`2341`)
```
```suggestion
* Ensure proper tz and pytz types in pvlib.location.Location.
```

Ensuring types is a bug fix. The remainder describes an API change that should be announced elsewhere. Not really an enhancement, so maybe we put the rest of this note in the "Breaking changes" section, because one can no longer set a value for `.pytz`.

markcampanelli replied:
@cwhanse Lmk if my `whatsnew` change for this looks ok.

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -31,6 +40,15 @@ Maintenance
 
+Breaking Changes
+~~~~~~~~~~~~~~~~
+* To ensure that the time zone in pvlib.location.Location remains internally
+  consistent if/when the time zone is updated, the tz attribute becomes the
+  single source of time-zone truth, is the single time-zone setter interface,
+  and its getter consistently returns an IANA string. datetime.tzinfo
+  subclasses (including time zones from pytz and zoneinfo) are still fully
+  supported in the tz setter, but the pytz attribute becomes read only.
+  (:issue:`2340`, :pull:`2341`)
```
```suggestion
* The pytz attribute of pvlib.location.Location is now read only. Use the tz attribute
  to record time zone. The pytz attribute is internally updated if the tz attribute
  is changed. The tz attribute 
* The tz attribute of pvlib.location.Location now returns an IANA string.
  (:issue:`2340`, :pull:`2341`)
```
I think the focus of this note should be on how the changes will affect user code.

markcampanelli replied:
I updated the whatsnew, trying to focus on the breaking change and its workaround, and also trying to provide the context behind the associated bugfix.

cwhanse replied:
Working on a second round of suggestions. The bug was that the docstring allowed for `tz` to be an int or float but inputs of those types weren't tested. The bug text in the last commit doesn't communicate that user code wasn't producing erroneous results.

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -31,6 +43,11 @@ Maintenance
 
+Breaking Changes
+~~~~~~~~~~~~~~~~
+* The pytz attribute in pvlib.location.Location is now read only. pytz time
+  zones can still be passed to tz in the object initializer and in the tz
+  setter after object creation. (:issue:`2340`, :pull:`2341`)
```
```suggestion
* The pvlib.location.Location.pytz attribute is now read only. The
  pytz attribute is now set internally to be consistent with the
  pvlib.location.Location.tz attribute. (:issue:`2340`, :pull:`2341`)
```
Maybe less technical would help.

cwhanse commented on pvlib/location.py:
```diff
@@ -38,12 +42,15 @@ class Location:
-        pytz.timezone objects will be converted to strings.
-        ints and floats must be in hours from UTC.
+    tz : time zone as str, int, float, or datetime.tzinfo, default 'UTC'.
+        This value represents as a valid IANA time zone name string. See
+        http://en.wikipedia.org/wiki/List_of_tz_database_time_zones for a
+        list of valid name strings, any of which may be passed directly here.
+        ints and floats must be whole-number hour offsets from UTC, which
+        are converted to the IANA-suppored 'Etc/GMT-N' format. (Note the
+        limited range of the offset N and its sign-change convention.)
+        Time zones from the pytz and zoneinfo packages may also be passed
+        directly here, as they are subclasses of datetime.tzinfo.
```
```suggestion
        See http://en.wikipedia.org/wiki/List_of_tz_database_time_zones for a
        list of valid name strings. An `int` or `float` must be a whole-number hour
        offsets from UTC, which can be converted to the IANA-supported
        'Etc/GMT-N' format. (Note the limited range of the offset N and its
        sign-change convention.) Time zones from the pytz and zoneinfo packages
        may also be passed here, as they are subclasses of datetime.tzinfo.
        
        The `tz` attribute is represented as a valid IANA time zone name string. 
```

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -12,12 +12,24 @@ Enhancements
 
 Testing
 ~~~~~~~
+* Add tests for timezone types in pvlib.location.Location.
```
```suggestion
* Add tests for all input types for the pvlib.location.Location.tz attribute.
```

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -12,12 +12,24 @@ Enhancements
 ~~~~~~~
+* Add tests for timezone types in pvlib.location.Location.
+  (:issue:`2340`, :pull:`2341`)
+* Add tests for time conversions in pvlib.tools. (:issue:`2340`, :pull:`2341`)
```
```suggestion
* Add tests for time conversion functions in pvlib.tools. (:issue:`2340`, :pull:`2341`)
```

cwhanse commented on docs/sphinx/source/whatsnew/v0.11.3.rst:
```diff
@@ -12,12 +12,24 @@ Enhancements
 
+Bug Fixes
+~~~~~~~~~
+* Ensure proper tz and pytz types in pvlib.location.Location. To ensure that
+  the time zone in pvlib.location.Location remains internally consistent
+  if/when the time zone is updated, the tz attribute is now the single source
+  of time-zone truth, is the single time-zone setter interface, and its getter 
+  returns an IANA string. (:issue:`2340`, :pull:`2341`)
```
```suggestion
* The pvlib.location.Location.tz attribute can be initialized with an `int` or
  a whole-number `float`. (:issue:`2340`, :pull:`2341`)
```
I really don't think this is the place to describe how the interface is changed.

kandersolar replied:
Seems to me like the "bug fix" here is not to do with what types are allowed, but rather that `Location` now ensures consistency between its `tz` and `pytz` attributes.

markcampanelli replied:
It's both. My memory is already fading, but I'm pretty sure a non-string value could be assigned to `pvlib.location.Location.tz`, and in addition to that this value and `pvlib.location.Location.pttz` could diverge.

markcampanelli replied:
My remaining concern here is to capture the reasoning behind a breaking interface change somewhere for posterity. If this can live somewhere else, I'm all for it.

kandersolar reviewed (approved):
This PR looks reasonable to me, although I am not very familiar with how `Location`'s timezone attributes are used throughout the library, or how we might expect users to use them for their own purposes.

cwhanse commented:
@markcampanelli let me know if you'd like help resolving the conflicts. We'd like to merge this for the March release.

markcampanelli commented:
@cwhanse I think I have addressed all remaining comments, except the level of detail provided when describing the bug fix. My remaining concern here is to capture somewhere for posterity the reasoning behind a breaking interface change that the bugfix engendered. If this can/should live somewhere else, I'm all for it.

cwhanse commented on pvlib/location.py:
```diff
@@ -54,43 +62,75 @@ class Location:
+    Raises
+    ------
+    ValueError
+        when the time-zone ``tz`` input cannot be converted.
```
```suggestion
        when the time zone ``tz`` cannot be converted.
```

cwhanse commented:
Thanks @markcampanelli for your perseverance.
